File: 322-coin-change.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        if (coins == null || coins.length == 0 || amount < 0) {
            return -1;
        }
        int[] dp = new int[amount + 1]; // dp[i] represents the min number of coin to get sum = i.
        dp[0] = 0;
        for (int i = 1; i < dp.length; i++) {
            int min = Integer.MAX_VALUE;
            for (int coin : coins) {
                if (i - coin >= 0 && dp[i - coin] != Integer.MAX_VALUE) {
                    min = Math.min(min, dp[i - coin]);
                }
            }
            if (min == Integer.MAX_VALUE) {
                dp[i] = Integer.MAX_VALUE;
            } else {
                dp[i] = min + 1;
            }
            
        }
        return dp[dp.length - 1] == Integer.MAX_VALUE ? -1 : dp[dp.length - 1];
    }
}

File: python3/322.coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        if amount == 0:
            return 0
        
        if not coins:
            return -1
        
        dp = [float('inf')] * (amount + 1)
        
        dp[0] = 0  
        
        for i in range(1, amount + 1):
            min_num = float('inf')
            for coin in coins:
                if i - coin >= 0:
                    
                    min_num = min(dp[i - coin], min_num)
                   
            dp[i] = min_num + 1        
         
        return dp[amount] if dp[amount] != float('inf') else -1
        
